File: src/runner.js
import BrythonRunner from 'brython-runner/lib/brython-runner.js'

import runnerPythonSource from './runner.py'

const MSG_TYPE_PREFIX = 'revealjs-ace-brython.'

class Runner {
  constructor() {
    this.createBrythonRunner()
    this.runSessionIdSeq = 0
    this.sessions = new Map()
    this.msgHandlers = {
      'exec-code-started': this.handleExecCodeStarted.bind(this),

      'exec-code-stdout-write': this.handleExecCodeStdoutWrite.bind(this),
      'exec-code-stdout-flush': this.handleExecCodeStdoutFlush.bind(this),
      'exec-code-stderr-write': this.handleExecCodeStderrWrite.bind(this),
      'exec-code-stderr-flush': this.handleExecCodeStderrFlush.bind(this),

      'exec-code-success': this.handleExecCodeSuccess.bind(this),
      'exec-code-error': this.handleExecCodeError.bind(this),

      'exec-success': this.handleExecSuccess.bind(this),
      'exec-error': this.handleExecError.bind(this),
    }
  }

  run(codeNodes, callbacks) {
    callbacks = {...callbacks}
    return new Promise((resolve, reject) => {
      const sessionId = this.runSessionIdSeq++
      const session = {
        codeNodes,
        callbacks,
        resolve,
        reject,
      }
      this.sessions.set(sessionId, session)
      const codes = codeNodes.map(n => n.code)
      this.sendMsg('exec', {session_id: sessionId, codes: codes})
    })
  }

  sendMsg(type, value) {
    type = `${MSG_TYPE_PREFIX}${type}`
    this.brythonRunner.sendMsg(type, value)
  }

  createBrythonRunner() {
    this.brythonRunner = new BrythonRunner({
      postInitScripts: [runnerPythonSource],
      // Use dummy stdout and stderr objects here because we handle output with
      // our own handlers.
      stdout: {write(content) {}, flush() {}},
      stderr: {write(content) {}, flush() {}},

      // Route messages with our prefix to our handlers
      onMsg: (type, value) => {
        if (!type.startsWith(MSG_TYPE_PREFIX)) {
          return
        }
        type = type.slice(MSG_TYPE_PREFIX.length)
        if (this.msgHandlers.hasOwnProperty(type)) {
          this.msgHandlers[type](value)
        } else {
          console.log(`Unhandled message of type '${type}'`, value)
        }
      },
    })
  }

  handleExecCodeStarted(data) {
    const session = this.sessions.get(data.session_id)
    if (!session.callbacks.codeStarted) {
      return
    }
    try {
      session.callbacks.codeStarted({codeIdx: data.code_idx})
    } catch {}
  }

  handleExecCodeStdoutWrite(data) {
    const session = this.sessions.get(data.session_id)
    if (!session.callbacks.stdout) {
      return
    }
    const codeIdx = data.code_idx
    try {
      session.callbacks.stdout({codeIdx: data.code_idx, data: data.data})
    } catch {}
  }

  handleExecCodeStdoutFlush(data) { /* Do nothing */ }

  handleExecCodeStderrWrite(data) {
    const session = this.sessions.get(data.session_id)
    if (!session.callbacks.stderr) {
      return
    }
    try {
      session.callbacks.stderr({codeIdx: data.code_idx, data: data.data})
    } catch {}
  }

  handleExecCodeStderrFlush(data) { /* Do nothing */ }

  handleExecCodeSuccess(data) {
    const session = this.sessions.get(data.session_id)
    if (!session.callbacks.codeSuccess) {
      return
    }
    try {
      session.callbacks.codeSuccess({codeIdx: data.code_idx})
    } catch {}
  }

  handleExecCodeError(data) {
    const session = this.sessions.get(data.session_id)
    if (!session.callbacks.codeError) {
      return
    }
    try {
      session.callbacks.codeError({codeIdx: data.code_idx, error: data.error})
    } catch {}
  }

  handleExecSuccess(data) {
    const session = this.sessions.get(data.session_id)
    session.resolve()
    this.sessions.delete(data.session_id)
  }

  handleExecError(data) {
    const session = this.sessions.get(data.session_id)
    session.reject(data.error)
    this.sessions.delete(data.session_id)
  }
}

export default Runner

File: src/runner.py
"""
Init script run by brython-runner for revealjs-ace-brython.
"""
import browser
import sys
import traceback

MSG_TYPE_PREFIX = 'revealjs-ace-brython.'


def send_msg(_type, value):
    """
    Send a message back to the main thread.
    """
    _type = f'{MSG_TYPE_PREFIX}{_type}'
    browser.self.send({'type': _type, 'value': value})


def handler(_type):
    """
    Decorator for defining a handler function.
    """
    def decorator(fn):
        handler.handlers[_type] = fn
        return fn
    return decorator
handler.handlers = {}


@browser.bind(browser.self, 'message')
def handle_message(evt):
    """
    Function that will handle a message from the main thread and route it to
    the appropriate handler.
    """
    if not evt.data.type.startswith(MSG_TYPE_PREFIX):
        return

    _type = evt.data.type[len(MSG_TYPE_PREFIX):]
    if _type not in handler.handlers:
        raise Exception(f'revealjs-ace-brython: no handler found for message type: {_type}')

    handler.handlers[_type](**evt.data.value)


class CustomOuputStream:
    def __init__(self, session_id, code_idx, parent_stream, msg_type_prefix):
        self.session_id = session_id
        self.code_idx = code_idx
        self.parent_stream = parent_stream
        self.msg_type_prefix = msg_type_prefix

    def write(self, data=''):
        send_msg(
            f'{self.msg_type_prefix}-write',
            {
                'session_id': self.session_id,
                'code_idx': self.code_idx,
                'data': str(data),
            },
        )
        self.parent_stream.write(data)

    def flush(self):
        send_msg(
            f'{self.msg_type_prefix}-flush',
            {
                'session_id': self.session_id,
                'code_idx': self.code_idx,
            },
        )
        self.parent_stream.flush()


@handler('exec')
def handle_exec(session_id, codes):
    g = {}
    send_msg('exec-started', {'session_id': session_id})
    saved_stdout = sys.stdout
    saved_stderr = sys.stderr
    try:
        for code_idx, code in enumerate(codes):
            send_msg(
                'exec-code-started',
                {'code_idx': code_idx, 'session_id': session_id},
            )

            sys.stdout = CustomOuputStream(
                session_id=session_id,
                code_idx=code_idx,
                parent_stream=saved_stdout,
                msg_type_prefix='exec-code-stdout',
            )

            sys.stderr = CustomOuputStream(
                session_id=session_id,
                code_idx=code_idx,
                parent_stream=saved_stderr,
                msg_type_prefix='exec-code-stderr',
            )

            try:
                exec(code, g)
            except:
                etype, evalue, tb = sys.exc_info()
                send_msg(
                    'exec-code-error',
                    {
                        'code_idx': code_idx,
                        'session_id': session_id,
                        'error': str(evalue),
                    },
                )
                traceback.print_exception(etype, evalue, tb)
                raise evalue
            else:
                send_msg(
                    'exec-code-success',
                    {'code_idx': code_idx, 'session_id': session_id},
                )
    except Exception as e:
        send_msg('exec-error', {'session_id': session_id, 'error': str(e)})
    else:
        send_msg('exec-success', {'session_id': session_id})
    finally:
        sys.stdout = saved_stdout
        sys.stderr = saved_stderr
